Shared (App)/Services/TransactionDecoder.swift

```diff
@@ -0,0 +1,42 @@
+//
+//  TransactionDecoder.swift
+//  Wallet
+//
+//  Created by Stefano on 21.12.21.
+//
+
+import Foundation
+import MEWwalletKit
+
+protocol TransactionDecodable {
+    func decodeInput(_ txInput: String, with contract: Contract) -> TransactionInput?
+}
+
+struct TransactionDecoder: TransactionDecodable {
+    
+    func decodeInput(_ txInput: String, with contract: Contract) -> TransactionInput? {
+        guard let data = contract.abi.data(using: .utf8),
+              let abiRecords = try? JSONDecoder().decode([ABI.Record].self, from: data) else {
+            return nil
+        }
+        for record in abiRecords {
+            let element = try? record.parse()
+            if case .function(let function) = element {
+                let signature = function.signature
+                let decodedMethodHash = function.methodString
+                let data = Data(hex: txInput)
+                guard !data.isEmpty, data.count > 4 else { return nil }
+                let methodHash = data[0 ..< 4].dataToHexString()
+                if methodHash == decodedMethodHash, let decodedInput = element?.decodeInputData(data) {
```

Comment:
@dimitarnestorov I'm currently using MewWalletKit to decode the tx inputs, which relies on ABIs. As discussed fetching the method Id from 4byte.directory would also work. But I think in this case I would probably need to implement a custom decoder (unless you know a Swift library that can decode the input based on the method signature only).

Reply:
I don't know of a library but I guess we can make one.

Reply:
I could start working on our own decoder but it will take me some time as I have limited availability in the next 1-2 weeks.

In regards to ABIs: would there be other benefits from having the entire ABI available for a particular contract? E.g. one thing we could do is to check whether a contract conforms to a particular interface (e.g. ERC721) and then tag or parse the tx detail accordingly.